contracts/utils/VaultQueue.sol

```diff
@@ -0,0 +1,223 @@
+            Vault.VaultParams memory vaultParams =
+                IRibbonVault(vault).vaultParams();
+            balance = IERC20(vaultParams.asset).balanceOf(address(this));
+        }
+    }
+
+    function withdrawToCreditor(
+        address vault,
+        address creditor,
+        uint256 portion
+    ) private {
+        if (vault == stethVault) {
+            address steth = IRibbonVault(vault).STETH();
+            IERC20(steth).transfer(creditor, portion);
+        } else if (vault == wethVault) {
+            payable(creditor).transfer(portion);
+        } else {
+            Vault.VaultParams memory vaultParams =
+                IRibbonVault(vault).vaultParams();
+            IERC20(vaultParams.asset).transfer(creditor, portion);
+        }
+    }
+
+    function transferToVault(
+        address depositContract,
+        address creditor,
+        uint256 portion
+    ) private {
+        IDepositContract(depositContract).depositFor{value: portion}(creditor);
+    }
+
+    function disburse(address vault) private {
+        uint256 balance = getInterVaultBalance(vault);
+        uint256 len = qTransfer[vault].length;
+        uint256 totalAmt = totalAmount[vault];
+        totalAmount[vault] = 0;
+
+        for (uint256 j = len; j > 0; j--) {
+            Transfer memory queue = pop(qTransfer[vault]);
+
+            uint256 portion =
+                DSMath.wmul(balance, DSMath.wdiv(queue.amount, totalAmt));
+            if (portion > 0) {
+                if (queue.action == TransferAction.INTERVAULT) {
+                    transferToVault(
+                        queue.depositContract,
+                        queue.creditor,
+                        portion
+                    );
+                } else if (queue.action == TransferAction.WITHDRAW) {
+                    withdrawToCreditor(vault, queue.creditor, portion);
+                }
+                emit Disburse(vault, queue, portion);
+            }
+        }
+    }
+
+    function transfer(address vault) external onlyOwner nonReentrant {
+        uint256 withdrawals = IRibbonVault(vault).withdrawals(address(this));
+        if (withdrawals > 0) {
+            if (vault == stethVault) {
+                IRibbonVault(vault).completeWithdraw(0);
+                disburse(vault);
+            } else {
+                IRibbonVault(vault).completeWithdraw();
+                disburse(vault);
+            }
+        }
+    }
+
+    function queueTransfer(
+        address srcVault,
+        address dstVault,
+        address depositContract,
+        TransferAction transferAction,
+        uint256 amount
+    ) external nonReentrant {
+        require(
+            !hasWithdrawal(srcVault, msg.sender),
+            "Withdraw already submitted"
+        );
+        require(qTransfer[srcVault].length < 256, "Transfer queue full");
+
+        if (transferAction == TransferAction.INTERVAULT) {
+            require(depositContract != address(0), "Intervault !address");
```

Comment:
Nit, don't think this revert message is helpful, maybe something like `No deposit contract`